src/MaxText/train.py

```diff
@@ -255,6 +255,8 @@ def reshape_to_microbatch_accumulations(batch_arr):
         extra_dpo_args = [reference_params]
     grad_func = jax.value_and_grad(_loss_fn, argnums=4, has_aux=True)
     (loss, aux), raw_grads = grad_func(model, config, data, dropout_rng, state.params, *extra_dpo_args, is_train=True)
+
+  raw_grads = jax.tree_util.tree_map(lambda x: x.astype(jnp.bfloat16) if x.dtype == jnp.float32 else x, raw_grads)
```

Comment:
Should this be based on config? Or is it always the case that this gets mapped back to `jnp.bfloat16`?

Reply:
nice catch, updated.

Reply:
Thanks @suexu1025. And are we also always checking for `x.dtype == jnp.float32`? Or can that be config-based too?

Reply:
yes, for here we only want `grad_dtype` to convert from fp32 to lower precision